File: src/main/java/io/femo/http/transport/http2/HttpFrameWriter.java
package io.femo.http.transport.http2;

import io.femo.http.Constants;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStream;

/**
 * Created by felix on 9/3/16.
 */
public class HttpFrameWriter {

    private static final Logger log = LoggerFactory.getLogger("HTTP/2.0");

    private OutputStream outputStream;
    private HttpConnection httpConnection;

    public HttpFrameWriter(OutputStream outputStream, HttpConnection httpConnection) {
        this.outputStream = outputStream;
        this.httpConnection = httpConnection;
    }

    public void write(HttpFrame httpFrame) throws IOException {
        try {
            httpFrame.build();
        } catch (Throwable t) {
            log.error("Error while building frame", t);
        }
        log.debug("Writing frame: " + httpFrame.toString());
        ByteArrayOutputStream outputBuffer = new ByteArrayOutputStream(httpFrame.getLength() + Constants.Http20.FRAME_HEADER_LENGTH);
        outputBuffer.write(HttpUtil.toByte(httpFrame.getLength()), 1, 3);
        outputBuffer.write(HttpUtil.toByte(httpFrame.getType()), 1, 1);
        outputBuffer.write(HttpUtil.toByte(httpFrame.getFlags()), 1, 1);
        outputBuffer.write(HttpUtil.toByte(httpFrame.getStreamIdentifier()));
        outputBuffer.write(httpFrame.getPayload());
        outputBuffer.writeTo(this.outputStream);
    }
}

File: src/main/java/org/web25/http/transport/http2/HttpFrameWriter.kt
package org.web25.http.transport.http2

import org.slf4j.LoggerFactory
import org.web25.http.Constants
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.OutputStream

/**
 * Created by felix on 9/3/16.
 */
class HttpFrameWriter(private val outputStream: OutputStream, private val httpConnection: HttpConnection) {

    @Throws(IOException::class)
    fun write(httpFrame: HttpFrame) {
        try {
            httpFrame.build()
        } catch (t: Throwable) {
            log.error("Error while building frame", t)
        }

        log.debug("Writing frame: " + httpFrame.toString())
        val outputBuffer = ByteArrayOutputStream(httpFrame.length + Constants.Http20.FRAME_HEADER_LENGTH)
        outputBuffer.write(HttpUtil.toByte(httpFrame.length), 1, 3)
        outputBuffer.write(HttpUtil.toByte(httpFrame.type), 1, 1)
        outputBuffer.write(HttpUtil.toByte(httpFrame.flags), 1, 1)
        outputBuffer.write(HttpUtil.toByte(httpFrame.streamIdentifier))
        outputBuffer.write(httpFrame.payload)
        outputBuffer.writeTo(this.outputStream)
    }

    companion object {

        private val log = LoggerFactory.getLogger("HTTP/2.0")
    }
}
